tensorstore/kvstore/s3/s3_key_value_store.cc

```diff
@@ -0,0 +1,1264 @@
+        credential_provider_ = nullptr;
+      } else {
+        TENSORSTORE_RETURN_IF_ERROR(result);
+        credential_provider_ = std::move(*result);
+      }
+    }
+    if(!*credential_provider_) return std::nullopt;
+    auto credential_result_ = (*credential_provider_)->GetCredentials();
+    if(!credential_result_.ok() && absl::IsNotFound(credential_result_.status())) {
+      return std::nullopt;
+    }
+    return credential_result_;
+  }
+
+// Apply default backoff/retry logic to the task.
+  // Returns whether the task will be retried. On false, max retries have
+  // been met or exceeded.  On true, `task->Retry()` will be scheduled to run
+  // after a suitable backoff period.
+  template <typename Task>
+  absl::Status BackoffForAttemptAsync(
+      absl::Status status, int attempt, Task* task,
+      SourceLocation loc = ::tensorstore::SourceLocation::current()) {
+    if (attempt >= spec_.retries->max_retries) {
+      return MaybeAnnotateStatus(
+          status,
+          tensorstore::StrCat("All ", attempt, " retry attempts failed"),
+          absl::StatusCode::kAborted, loc);
+    }
+
+    // https://cloud.google.com/storage/docs/retry-strategy#exponential-backoff
+    s3_retries.Increment();
+    auto delay = internal::BackoffForAttempt(
+        attempt, spec_.retries->initial_delay, spec_.retries->max_delay,
+        /*jitter=*/std::min(absl::Seconds(1), spec_.retries->initial_delay));
+        ScheduleAt(absl::Now() + delay,
+               WithExecutor(executor(), [task = IntrusivePtr<Task>(task)] {
+                 task->Retry();
+               }));
+
+    return absl::OkStatus();
+  }
+
+
+  std::shared_ptr<HttpTransport> transport_;
+  internal_storage_gcs::NoRateLimiter no_rate_limiter_;
+  std::string endpoint_;  // endpoint url
+  SpecData spec_;
+
+  absl::Mutex credential_provider_mutex_;
+  std::optional<std::shared_ptr<CredentialProvider>> credential_provider_;
+  std::string aws_region_;
+};
+
+
+Future<kvstore::DriverPtr> S3KeyValueStoreSpec::DoOpen() const {
+  auto driver = internal::MakeIntrusivePtr<S3KeyValueStore>();
+  driver->spec_ = data_;
+  driver->transport_ = internal_http::GetDefaultHttpTransport();
+
+  if(absl::StripAsciiWhitespace(data_.endpoint).empty()) {
+    // Assume AWS
+    // Make global request to get bucket region from response headers,
+    // then create region specific endpoint
+    auto url = tensorstore::StrCat("https://", data_.bucket, ".s3", kDotAmazonAwsDotCom);
+    auto future = driver->transport_->IssueRequest(
+            HttpRequestBuilder("GET", url).BuildRequest(),
+            absl::Cord());
+    if(!future.status().ok()) return future.status();
+    auto & headers = future.value().headers;
+    if(auto it = headers.find(kAmzBucketRegionHeader); it !=headers.end()) {
+      driver->aws_region_ = it->second;
+      driver->endpoint_ = tensorstore::StrCat(
+        "http://", data_.bucket, ".s3.", driver->aws_region_, kDotAmazonAwsDotCom);
```

Comment:
This should use `https` instead of `http`.

Reply:
Is there any advantage in using the `<bucket>.s3.<region>.amazonaws.com` endpoint rather than just `<bucket>.s3.amazonaws.com`?  Since `<bucket>.s3.amazonaws.com` hostname is used to determine the location, I expect it would be more efficient to just continue using that hostname.